pkg/cmd/bazci/bazci.go

```diff
@@ -283,59 +279,34 @@ func init() {
 	)
 }
 
-func bazciImpl(cmd *cobra.Command, args []string) (retErr error) {
-	var goTestJSONOutputFilePath string
-	defer func() {
-		if err := processTestJSONIfNeeded(retErr != nil /* shouldCreateTarball */, goTestJSONOutputFilePath); err != nil {
-			fmt.Printf("failed to process go test json output - %v\n", err)
-		}
-	}()
-
+func bazciImpl(cmd *cobra.Command, args []string) error {
 	if args[0] != buildSubcmd && args[0] != runSubcmd && args[0] != testSubcmd && args[0] != mungeTestXMLSubcmd && args[0] != mergeTestXMLsSubcmd {
-		retErr = errors.Newf("First argument must be `build`, `run`, `test`, `merge-test-xmls`, or `munge-test-xml`; got %v", args[0])
-		return
+		return errors.Newf("First argument must be `build`, `run`, `test`, `merge-test-xmls`, or `munge-test-xml`; got %v", args[0])
 	}
 
 	// Special case: munge-test-xml/merge-test-xmls don't require running Bazel at all.
 	// Perform the munge then exit immediately.
 	if args[0] == mungeTestXMLSubcmd {
-		retErr = mungeTestXMLs(args)
-		return
+		return mungeTestXMLs(args)
 	}
 	if args[0] == mergeTestXMLsSubcmd {
-		retErr = mergeTestXMLs(args)
-		return
+		return mergeTestXMLs(args)
 	}
 
 	server := newMonitorBuildServer(args[0])
 	if err := server.Start(); err != nil {
-		retErr = err
-		return
+		return err
 	}
 	args = append(args, fmt.Sprintf("--bes_backend=grpc://127.0.0.1:%d", port))
-	if shouldProcessTestFailures {
-		f, createTempErr := os.CreateTemp(artifactsDir, "test.json.txt")
-		if createTempErr != nil {
-			retErr = createTempErr
-			return
-		}
-		goTestJSONOutputFilePath = f.Name()
-		// Closing the file because we will not use the file pointer.
-		if retErr = f.Close(); retErr != nil {
-			return
-		}
-		args = append(args, "--test_env", goTestJSONOutputFilePath)
-	}
 	fmt.Println("running bazel w/ args: ", shellescape.QuoteCommand(args))
 	bazelCmd := exec.Command("bazel", args...)
 	bazelCmd.Stdout = os.Stdout
 	bazelCmd.Stderr = os.Stderr
-	retErr = bazelCmd.Run()
-	if retErr != nil {
-		fmt.Printf("got error %+v from bazel run\n", retErr)
+	if err := bazelCmd.Run(); err != nil {
+		fmt.Printf("got error %+v from bazel run\n", err)
 	}
 	server.Wait()
-	return
+	return processTestXmls(server.testXmls)
```

Comment:
Nice work getting rid of `retErr`, is it safe to assume that `processTestXmls` will always be non-nil if `bazelCmd.Run()` is non-nil? An example in mind is if `bazel` fails to start any tests due to a configuration issue we may not have any `test.xml` file so `processTestXmls` would be nil and the TeamCity run will be green. I am not sure how true this is, but if it makes sense we may want to check both and return an error if any of them is non-nil (after calling `processTestXmls`).

Reply:
Fixed, have a look.